### PR TITLE
Add a flag to get amount results as integer

### PR DESCRIPTION
**Pull request by oxarbitrage**
Closes https://github.com/zcash/zcash/issues/2375

### CODE REVIEW

**Comment by oxarbitrage** [comment]
Bellow is a list of modified rpc methods where amount is returned inside `rpc_wallet.cpp`. 
4 methods need an optional boolean argument `inZat` to get the amount in zatoshis, this is needed as these functions currently only return a number. In the rest `amountZat` field was added everywhere after `amount` output.

- `getreceivedbyaddress` - bool arg `inZat` added.
- `getreceivedbyaccount` - bool arg `inZat` added.
- `getbalance` - bool arg `inZat` added.
- `listreceivedbyaddress` - output field `amountZat` added.
- `listreceivedbyaccount` - output field `amountZat` added.
- `listtransactions` - output field `amountZat` added.
- `listsinceblock` - output field `amountZat` added.
- `gettransaction` - output field `amountZat` added.
- `listunspent` - output field `amountZat` added.
- `z_listreceivedbyaddress` - output field `amountZat` added.
- `z_getbalance` - bool arg `inZat` added.

File: qa/rpc-tests/wallet.py [hunk 1]
```diff
@@ -381,15 +388,18 @@ def run_test (self):
...
         txId  = self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), "0.0001")
         txObj = self.nodes[0].gettransaction(txId)
         assert_equal(txObj['amount'], Decimal('-0.00010000'))
+        assert_equal(txObj['amountZat'], int('-10000'))
 
         #check if JSON parser can handle scientific notation in strings
         txId  = self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), "1e-4")
         txObj = self.nodes[0].gettransaction(txId)
         assert_equal(txObj['amount'], Decimal('-0.00010000'))
+        assert_equal(txObj['amountZat'], int('-10000'))
```
Comment (daira) [inline qa/rpc-tests/wallet.py:402]:
```suggestion
        assert_equal(txObj['amountZat'], -10000)
```

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -633,7 +634,12 @@ UniValue getreceivedbyaddress(const UniValue& params, bool fHelp)
                     nAmount += txout.nValue;
     }
 
-    return  ValueFromAmount(nAmount);
+    // inZat
+    if (params.size() > 2)
+        if (params[2].get_bool())
+            return nAmount;
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:640]:
```suggestion
    if (params.size() > 2 && params[2].get_bool()) {
        return nAmount;
    }
```

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -690,6 +697,11 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
         }
     }
 
+    // inZat
+    if (params.size() > 2)
+        if (params[2].get_bool())
+            return nAmount;
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:703]:
```suggestion
    if (params.size() > 2 && params[2].get_bool()) {
        return nAmount;
    }
```

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -788,6 +801,12 @@ UniValue getbalance(const UniValue& params, bool fHelp)
                 nBalance -= s.amount;
             nBalance -= allFee;
         }
+
+        // inZat
+        if (params.size() > 3)
+            if (params[3].get_bool())
+                return nBalance;
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:808]:
```suggestion
        if (params.size() > 3 && params[3].get_bool()) {
            return nAmount;
        }
```

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -1444,6 +1470,7 @@ UniValue listtransactions(const UniValue& params, bool fHelp)
             "    \"amount\": x.xxx,          (numeric) The amount in " + CURRENCY_UNIT + ". This is negative for the 'send' category, and for the\n"
             "                                         'move' category for moves outbound. It is positive for the 'receive' category,\n"
             "                                         and for the 'move' category for inbound funds.\n"
+            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as 'amount' field.\n"
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:1473]:
```suggestion
            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as for the 'amount' field.\n"
```

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -1639,6 +1666,7 @@ UniValue listsinceblock(const UniValue& params, bool fHelp)
             "    \"category\":\"send|receive\",     (string) The transaction category. 'send' has negative amounts, 'receive' has positive amounts.\n"
             "    \"amount\": x.xxx,          (numeric) The amount in " + CURRENCY_UNIT + ". This is negative for the 'send' category, and for the 'move' category for moves \n"
             "                                          outbound. It is positive for the 'receive' category, and for the 'move' category for inbound funds.\n"
+            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as 'amount' field.\n"
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:1669]:
```suggestion
            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as for the 'amount' field.\n"
```

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -1268,6 +1289,7 @@ UniValue listreceivedbyaddress(const UniValue& params, bool fHelp)
             "    \"address\" : \"receivingaddress\",  (string) The receiving address\n"
             "    \"account\" : \"accountname\",       (string) DEPRECATED. The account of the receiving address. The default account is \"\".\n"
             "    \"amount\" : x.xxx,                  (numeric) The total amount in " + CURRENCY_UNIT + " received by the address\n"
+            "    \"amountZat\" : xxxx                 (numeric) The amount in zatoshis\n"
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:1292]: Add `MINOR_CURRENCY_UNIT` set to "zatoshis", analogous to `CURRENCY_UNIT`. (The "Zat" suffixes on field names don't need to change.) This could be a separate PR if you prefer.

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -3461,6 +3498,11 @@ UniValue z_getbalance(const UniValue& params, bool fHelp)
         nBalance = getBalanceZaddr(fromaddress, nMinDepth, false);
     }
 
+    // inZat
+    if (params.size() > 3)
+        if (params[3].get_bool())
+            return nBalance;
```
Comment (daira) [inline src/wallet/rpcwallet.cpp:3504]:
```suggestion
    if (params.size() > 3 && params[3].get_bool()) {
        return nBalance;
    }
```

**Comment by oxarbitrage** [comment]
Please don't merge, working in the suggested changes.

**Comment by oxarbitrage** [comment]
the comments and suggestions are all done in the last 5 commits.

File: qa/rpc-tests/receivedby.py [hunk 9]
```diff
@@ -73,19 +74,19 @@ def run_test(self):
         self.sync_all()
         check_array_result(self.nodes[1].listreceivedbyaddress(),
                            {"address":addr},
-                           {"address":addr, "account":"", "amount":Decimal("0.1"), "confirmations":10, "txids":[txid,]})
+                           {"address":addr, "account":"", "amount":Decimal("0.1"), "amountZat":int("10000000"), "confirmations":10, "txids":[txid,]})
```
Comment (daira) [inline qa/rpc-tests/receivedby.py:77]: The integers in this file don't need `int()`. (Does not block.)

File: qa/rpc-tests/wallet.py [hunk 10]
```diff
@@ -222,6 +222,7 @@ def run_test (self):
             if uTx['txid'] == zeroValueTxid:
                 found = True
                 assert_equal(uTx['amount'], Decimal('0.00000000'))
+                assert_equal(uTx['amountZat'], int('0'))
```
Comment (daira) [inline qa/rpc-tests/wallet.py:225]: Same here, no need for `int()`.

File: qa/rpc-tests/wallet_listreceived.py [hunk 11]
```diff
@@ -16,6 +16,7 @@
 no_memo = 'f6' + ('0'*1022) # see section 5.5 of the protocol spec
 
 fee = Decimal('0.0001')
+feeZat = Decimal('10000')
```
Comment (daira) [inline qa/rpc-tests/wallet_listreceived.py:19]: No need for `Decimal` here, just use integers.

File: qa/rpc-tests/wallet_listreceived.py [hunk 12]
```diff
@@ -71,11 +73,13 @@ def run_test_release(self, release, height):
 
         assert_equal(txid, r[0]['txid'])
         assert_equal(Decimal('0.4')-fee, r[0]['amount'])
+        assert_equal(Decimal('40000000')-feeZat, r[0]['amountZat'])
```
Comment (daira) [inline qa/rpc-tests/wallet_listreceived.py:76]: and here.

**Comment by oxarbitrage** [comment]
I rebased and added a new commit to fix the last comments. Let me know how do you see it. Thank you @daira

File: qa/rpc-tests/wallet.py [hunk 13]
```diff
@@ -263,10 +264,13 @@ def run_test (self):
         self.nodes[0].generate(1)
         sync_blocks(self.nodes)
 
-        #tx should be added to balance because after restarting the nodes tx should be broadcastet
+        # tx should be added to balance because after restarting the nodes tx should be broadcast
         assert_equal(self.nodes[2].getbalance(), Decimal('13.99800000')) #should not be
         assert_equal(self.nodes[2].getbalance("*"), Decimal('13.99800000')) #should not be
 
+        # check integer balances from getbalance
+        assert_equal(self.nodes[2].getbalance("*", 1, False, True), 1399800000) #should not be
```
Comment (daira) [inline qa/rpc-tests/wallet.py:272]: Nonblocking: does anyone know what "should not be?" means in these three comments?

Comment (daira) [reply qa/rpc-tests/wallet.py:272]: Seems like it has been there since the code was added in https://github.com/zcash/zcash/commit/77650cc9f4646787e843b2922730d8cba5af36a1 . I suggest just removing it (on 3 lines).

Comment (oxarbitrage) [reply qa/rpc-tests/wallet.py]: Removed comments at https://github.com/zcash/zcash/pull/4282/commits/76cb4e8d0437be9f2843d7affac0f7573605dd12

File: qa/rpc-tests/receivedby.py [hunk 14]
```diff
@@ -71,19 +72,19 @@ def run_test(self):
...
+                           {"address":addr, "account":"", "amount":Decimal("0.1"), "amountZat":10000000, "confirmations":10, "txids":[txid,]})
         # With min confidence > 10, should not find Tx
         check_array_result(self.nodes[1].listreceivedbyaddress(11),{"address":addr},{ },True)
 
         # Empty Tx
         addr = self.nodes[1].getnewaddress()
         check_array_result(self.nodes[1].listreceivedbyaddress(0,True),
                            {"address":addr},
-                           {"address":addr, "account":"", "amount":0, "confirmations":0, "txids":[]})
+                           {"address":addr, "account":"", "amount":0, "confirmations":0, "amountZat":int("0"), "txids":[]})
```
Comment (str4d) [inline qa/rpc-tests/receivedby.py:87]:
```suggestion
                           {"address":addr, "account":"", "amount":0, "confirmations":0, "amountZat":0, "txids":[]})
```

File: src/wallet/rpcwallet.cpp [hunk 15]
```diff
@@ -597,13 +597,14 @@ UniValue getreceivedbyaddress(const UniValue& params, bool fHelp)
...
+    if (fHelp || params.size() < 1 || params.size() > 3)
         throw runtime_error(
-            "getreceivedbyaddress \"zcashaddress\" ( minconf )\n"
+            "getreceivedbyaddress \"zcashaddress\" ( minconf ) ( inZat )\n"
             "\nReturns the total amount received by the given Zcash address in transactions with at least minconf confirmations.\n"
             "\nArguments:\n"
             "1. \"zcashaddress\"  (string, required) The Zcash address for transactions.\n"
-            "2. minconf             (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "2. minconf         (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:607]:
```suggestion
            "3. inZat           (bool, optional, default=false) Get the result amount in " + MINOR_CURRENCY_UNIT + " (as an integer).\n"
```

File: src/wallet/rpcwallet.cpp [hunk 16]
```diff
@@ -657,13 +663,14 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
...
+    if (fHelp || params.size() < 1 || params.size() > 3)
         throw runtime_error(
-            "getreceivedbyaccount \"account\" ( minconf )\n"
+            "getreceivedbyaccount \"account\" ( minconf ) ( inZat )\n"
             "\nDEPRECATED. Returns the total amount received by addresses with <account> in transactions with at least [minconf] confirmations.\n"
             "\nArguments:\n"
             "1. \"account\"      (string, required) MUST be set to the empty string \"\" to represent the default account. Passing any other string will result in an error.\n"
-            "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "2. minconf        (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:673]: Ditto.

File: src/wallet/rpcwallet.cpp [hunk 17]
```diff
@@ -746,14 +758,15 @@ UniValue getbalance(const UniValue& params, bool fHelp)
...
+    if (fHelp || params.size() > 4)
         throw runtime_error(
-            "getbalance ( \"account\" minconf includeWatchonly )\n"
+            "getbalance ( \"account\" minconf includeWatchonly inZat )\n"
             "\nReturns the server's total available balance.\n"
             "\nArguments:\n"
             "1. \"account\"      (string, optional) DEPRECATED. If provided, it MUST be set to the empty string \"\" or to the string \"*\", either of which will give the total available balance. Passing any other string will result in an error.\n"
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
             "3. includeWatchonly (bool, optional, default=false) Also include balance in watchonly addresses (see 'importaddress')\n"
+            "4. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:769]: Ditto.

File: src/wallet/rpcwallet.cpp [hunk 18]
```diff
@@ -3451,15 +3487,16 @@ UniValue z_getbalance(const UniValue& params, bool fHelp)
...
         throw runtime_error(
-            "z_getbalance \"address\" ( minconf )\n"
+            "z_getbalance \"address\" ( minconf inZat )\n"
             "\nReturns the balance of a taddr or zaddr belonging to the node's wallet.\n"
             "\nCAUTION: If the wallet has only an incoming viewing key for this address, then spends cannot be"
             "\ndetected, and so the returned balance may be larger than the actual balance.\n"
             "\nArguments:\n"
             "1. \"address\"      (string) The selected address. It may be a transparent or private address.\n"
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:3499]: Ditto.

File: src/wallet/rpcwallet.cpp [hunk 19]
```diff
@@ -597,13 +597,14 @@ UniValue getreceivedbyaddress(const UniValue& params, bool fHelp)
...
-            "getreceivedbyaddress \"zcashaddress\" ( minconf )\n"
+            "getreceivedbyaddress \"zcashaddress\" ( minconf ) ( inZat )\n"
             "\nReturns the total amount received by the given Zcash address in transactions with at least minconf confirmations.\n"
             "\nArguments:\n"
             "1. \"zcashaddress\"  (string, required) The Zcash address for transactions.\n"
-            "2. minconf             (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "2. minconf         (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount   (numeric) The total amount in " + CURRENCY_UNIT + " received at this address.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:609]: This needs to be updated to reflect the units variability.

File: src/wallet/rpcwallet.cpp [hunk 20]
```diff
@@ -657,13 +663,14 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
...
-            "getreceivedbyaccount \"account\" ( minconf )\n"
+            "getreceivedbyaccount \"account\" ( minconf ) ( inZat )\n"
             "\nDEPRECATED. Returns the total amount received by addresses with <account> in transactions with at least [minconf] confirmations.\n"
             "\nArguments:\n"
             "1. \"account\"      (string, required) MUST be set to the empty string \"\" to represent the default account. Passing any other string will result in an error.\n"
-            "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "2. minconf        (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:675]: Ditto.

File: src/wallet/rpcwallet.cpp [hunk 21]
```diff
@@ -746,14 +758,15 @@ UniValue getbalance(const UniValue& params, bool fHelp)
...
-            "getbalance ( \"account\" minconf includeWatchonly )\n"
+            "getbalance ( \"account\" minconf includeWatchonly inZat )\n"
             "\nReturns the server's total available balance.\n"
             "\nArguments:\n"
             "1. \"account\"      (string, optional) DEPRECATED. If provided, it MUST be set to the empty string \"\" or to the string \"*\", either of which will give the total available balance. Passing any other string will result in an error.\n"
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
             "3. includeWatchonly (bool, optional, default=false) Also include balance in watchonly addresses (see 'importaddress')\n"
+            "4. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:771]: Ditto.

File: src/wallet/rpcwallet.cpp [hunk 22]
```diff
@@ -3451,15 +3487,16 @@ UniValue z_getbalance(const UniValue& params, bool fHelp)
...
+            "z_getbalance \"address\" ( minconf inZat )\n"
             "\nReturns the balance of a taddr or zaddr belonging to the node's wallet.\n"
             "\nCAUTION: If the wallet has only an incoming viewing key for this address, then spends cannot be"
             "\ndetected, and so the returned balance may be larger than the actual balance.\n"
             "\nArguments:\n"
             "1. \"address\"      (string) The selected address. It may be a transparent or private address.\n"
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this address.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:3501]: Ditto.

**Comment by oxarbitrage** [comment]
Add the parameters to `src/rpc/client.cpp` is needed otherwise will not work for the cli. Thank you @str4d . I added them at https://github.com/zcash/zcash/pull/4282/commits/93afebeef43597610ef8bad66c1ae1c9e69926b7 

While testing that i found some errors that were also fixed in the same commit.

File: src/wallet/rpcwallet.cpp [hunk 23]
```diff
@@ -670,9 +670,9 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
             "\nArguments:\n"
             "1. \"account\"      (string, required) MUST be set to the empty string \"\" to represent the default account. Passing any other string will result in an error.\n"
             "2. minconf        (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
-            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount in " + MINOR_CURRENCY_UNIT + " (as an integer).\n"
             "\nResult:\n"
-            "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
+            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received at this address.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:675]: Copy-paste error:
```suggestion
            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received for this account.\n"
```

File: src/wallet/rpcwallet.cpp [hunk 24]
```diff
@@ -766,9 +766,9 @@ UniValue getbalance(const UniValue& params, bool fHelp)
             "1. \"account\"      (string, optional) DEPRECATED. If provided, it MUST be set to the empty string \"\" or to the string \"*\", either of which will give the total available balance. Passing any other string will result in an error.\n"
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
             "3. includeWatchonly (bool, optional, default=false) Also include balance in watchonly addresses (see 'importaddress')\n"
-            "4. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
+            "4. inZat            (bool, optional, default=false) Get the result amount in " + MINOR_CURRENCY_UNIT + " (as an integer).\n"
             "\nResult:\n"
-            "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
+            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received at this address.\n"
```
Comment (str4d) [inline src/wallet/rpcwallet.cpp:771]: Copy-paste error, but here we can improve the message (since accounts are deprecated):
```suggestion
            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received.\n"
```

**Comment by oxarbitrage** [comment]
Thank you @str4d . I used the github "add to batch" feature to include your suggestions. Hope is ok.

**Comment by str4d** [comment]
@zkbot r+

File: qa/rpc-tests/listtransactions.py [hunk 25]
```diff
@@ -6,6 +6,7 @@
 # Exercise the listtransactions API
 
 from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
```
Comment (rex4539) [inline qa/rpc-tests/listtransactions.py:9]:
```suggestion
```

**Comment by daira** [comment]
utACK. @zkbot r+
